promql/engine.go

```diff
@@ -606,88 +579,73 @@ func (ev *evaluator) recover(errp *error) {
-func (ev *evaluator) evalVector(e Expr) Vector {
-	val := ev.eval(e)
-	vec, ok := val.(Vector)
-	if !ok {
-		ev.errorf("expected instant Vector but got %s", documentedType(val.Type()))
-	}
-	return vec
-}
+// Extra information and caches for evaluating a single node across steps.
+type evalNodeHelper struct {
+	// Evaluation timestamp.
+	ts int64
+	// Vector that can be used for output.
+	out Vector
 
-// evalInt attempts to evaluate e into an integer and errors otherwise.
-func (ev *evaluator) evalInt(e Expr) int64 {
-	sc := ev.evalScalar(e)
-	if !convertibleToInt64(sc.V) {
-		ev.errorf("Scalar value %v overflows int64", sc.V)
-	}
-	return int64(sc.V)
-}
+	// Caches.
+	dmn                          map[uint64]labels.Labels      // dropMetricName and label_*.
+	sigf                         map[uint64]uint64             // signatureFunc.
+	signatureToMetricWithBuckets map[uint64]*metricWithBuckets // funcHistogramQuantile.
+	regex                        *regexp.Regexp                // label_replace.
 
-// evalFloat attempts to evaluate e into a float and errors otherwise.
-func (ev *evaluator) evalFloat(e Expr) float64 {
-	sc := ev.evalScalar(e)
-	return float64(sc.V)
+	// For binary vector matching.
+	rightSigs    map[uint64]Sample
+	matchedSigs  map[uint64]map[uint64]struct{}
+	resultMetric map[uint64]labels.Labels
 }
 
-// evalMatrix attempts to evaluate e into a Matrix and errors otherwise.
-// The error message uses the term "range Vector" to match the user facing
-// documentation.
-func (ev *evaluator) evalMatrix(e Expr) Matrix {
-	val := ev.eval(e)
-	mat, ok := val.(Matrix)
-	if !ok {
-		ev.errorf("expected range Vector but got %s", documentedType(val.Type()))
+// dropMetricName is a cached version of dropMetricName.
+func (enh *evalNodeHelper) dropMetricName(l labels.Labels) labels.Labels {
+	if enh.dmn == nil {
+		enh.dmn = make(map[uint64]labels.Labels, len(enh.out))
 	}
-	return mat
-}
-
-// evalString attempts to evaluate e to a string value and errors otherwise.
-func (ev *evaluator) evalString(e Expr) String {
-	val := ev.eval(e)
-	sv, ok := val.(String)
-	if !ok {
-		ev.errorf("expected string but got %s", documentedType(val.Type()))
+	h := l.Hash()
+	ret, ok := enh.dmn[h]
+	if ok {
+		return ret
 	}
-	return sv
+	ret = dropMetricName(l)
+	enh.dmn[h] = ret
+	return ret
 }
 
-// evalOneOf evaluates e and errors unless the result is of one of the given types.
-func (ev *evaluator) evalOneOf(e Expr, t1, t2 ValueType) Value {
-	val := ev.eval(e)
-	if val.Type() != t1 && val.Type() != t2 {
-		ev.errorf("expected %s or %s but got %s", documentedType(t1), documentedType(t2), documentedType(val.Type()))
+// signatureFunc is a cached version of signatureFunc.
+func (enh *evalNodeHelper) signatureFunc(on bool, names ...string) func(labels.Labels) uint64 {
+	if enh.sigf == nil {
+		enh.sigf = make(map[uint64]uint64, len(enh.out))
+	}
+	f := signatureFunc(on, names...)
```

Comment:
Does it generally still make sense to have the methods attached to the helper _and_ the normal functions. Seems a bit confusing and I could see people using the plain one if they modify code somewhere.

Reply:
We'd hopefully catch it with benchmarks. This might be something that can be inlined later on with future optimisations I have in mind.

Reply:
"Hope is not a strategy." *-- Google SRE*